File: frontend/src/api/fluidEntries.ts
import { API_BASE_URL } from "./config";
import { responseError } from "./errors";

export type FluidEntry = {
  id: number;
  entry_time: string;
  amount_ml: number;
  notes: string | null;
  created_at: string;
};

export type FluidEntryCreate = {
  entry_time?: string;
  amount_ml: number;
  notes?: string | null;
};

export type FluidEntryUpdate = {
  entry_time: string;
  amount_ml: number;
  notes?: string | null;
};

export async function getFluidEntries(): Promise<FluidEntry[]> {
  const response = await fetch(`${API_BASE_URL}/fluid-entries`);

  if (!response.ok) {
    throw new Error("Could not load fluid entries.");
  }

  return response.json();
}

export async function createFluidEntry(
  fluidEntry: FluidEntryCreate,
): Promise<FluidEntry> {
  const response = await fetch(`${API_BASE_URL}/fluid-entries`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(fluidEntry),
  });

  if (!response.ok) {
    throw new Error("Could not create fluid entry.");
  }

  return response.json();
}

export async function updateFluidEntry(
  fluidEntryId: number,
  fluidEntry: FluidEntryUpdate,
): Promise<FluidEntry> {
  const response = await fetch(`${API_BASE_URL}/fluid-entries/${fluidEntryId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(fluidEntry),
  });

  if (!response.ok) {
    throw await responseError(response, "Could not update fluid entry.");
  }

  return response.json();
}

export async function deleteFluidEntry(fluidEntryId: number): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/fluid-entries/${fluidEntryId}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error("Could not delete fluid entry.");
  }
}

File: backend/app/routes/fluid_entries.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app import models, schemas
from app.database import SessionLocal
from app.time_utils import entry_time_or_now

router = APIRouter(prefix="/fluid-entries", tags=["Fluid Entries"])


def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.post(
    "",
    response_model=schemas.FluidEntryResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_fluid_entry(
    fluid_entry: schemas.FluidEntryCreate,
    db: Session = Depends(get_db),
):
    db_fluid_entry = models.FluidEntry(
        entry_time=entry_time_or_now(fluid_entry.entry_time),
        amount_ml=fluid_entry.amount_ml,
        notes=fluid_entry.notes,
    )

    db.add(db_fluid_entry)
    db.commit()
    db.refresh(db_fluid_entry)

    return db_fluid_entry


@router.get("", response_model=list[schemas.FluidEntryResponse])
def list_fluid_entries(db: Session = Depends(get_db)):
    return (
        db.query(models.FluidEntry)
        .order_by(models.FluidEntry.entry_time.desc(), models.FluidEntry.id.desc())
        .all()
    )


@router.patch("/{fluid_entry_id}", response_model=schemas.FluidEntryResponse)
def update_fluid_entry(
    fluid_entry_id: int,
    fluid_entry_update: schemas.FluidEntryUpdate,
    db: Session = Depends(get_db),
):
    db_fluid_entry = (
        db.query(models.FluidEntry)
        .filter(models.FluidEntry.id == fluid_entry_id)
        .first()
    )

    if db_fluid_entry is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Fluid entry not found",
        )

    db_fluid_entry.entry_time = entry_time_or_now(fluid_entry_update.entry_time)
    db_fluid_entry.amount_ml = fluid_entry_update.amount_ml
    db_fluid_entry.notes = fluid_entry_update.notes

    db.commit()
    db.refresh(db_fluid_entry)

    return db_fluid_entry


@router.delete("/{fluid_entry_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_fluid_entry(fluid_entry_id: int, db: Session = Depends(get_db)):
    db_fluid_entry = (
        db.query(models.FluidEntry)
        .filter(models.FluidEntry.id == fluid_entry_id)
        .first()
    )

    if db_fluid_entry is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Fluid entry not found",
        )

    db.delete(db_fluid_entry)
    db.commit()
